File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/CookieBuilder.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws.ahc

import play.api.libs.ws.DefaultWSCookie
import play.api.libs.ws.WSCookie
import play.shaded.ahc.io.netty.handler.codec.http.HttpHeaderNames._
import play.shaded.ahc.io.netty.handler.codec.http.cookie.ClientCookieDecoder
import play.shaded.ahc.io.netty.handler.codec.http.cookie.Cookie
import play.shaded.ahc.io.netty.handler.codec.http.cookie.DefaultCookie

trait CookieBuilder extends WSCookieConverter {
  def buildCookies(headers: Map[String, scala.collection.Seq[String]]): scala.collection.Seq[WSCookie] = {
    val option = headers.get(SET_COOKIE2.toString).orElse(headers.get(SET_COOKIE.toString))
    option
      .map { cookiesHeaders =>
        for {
          value <- cookiesHeaders
          Some(c) = Some(
            if (useLaxCookieEncoder) ClientCookieDecoder.LAX.decode(value)
            else ClientCookieDecoder.STRICT.decode(value)
          )
        } yield asCookie(c)
      }
      .getOrElse(Seq.empty)
  }

  def useLaxCookieEncoder: Boolean
}

/**
 * Converts between AHC cookie and the WS cookie.
 */
trait WSCookieConverter {

  def asCookie(cookie: WSCookie): Cookie = {
    val c = new DefaultCookie(cookie.name, cookie.value)
    c.setWrap(false)
    c.setDomain(cookie.domain.orNull)
    c.setPath(cookie.path.orNull)
    c.setMaxAge(cookie.maxAge.getOrElse(-1L))
    c.setSecure(cookie.secure)
    c.setHttpOnly(cookie.httpOnly)
    c
  }

  def asCookie(c: Cookie): WSCookie = {
    DefaultWSCookie(
      name = c.name,
      value = c.value,
      domain = Option(c.domain),
      path = Option(c.path),
      maxAge = Option(c.maxAge).filterNot(_ < 0),
      secure = c.isSecure,
      httpOnly = c.isHttpOnly
    )
  }
}

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/CookieBuilder.java
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws.ahc;

import play.libs.ws.WSCookie;
import play.libs.ws.WSCookieBuilder;
import play.shaded.ahc.io.netty.handler.codec.http.cookie.ClientCookieDecoder;
import play.shaded.ahc.io.netty.handler.codec.http.cookie.Cookie;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static play.shaded.ahc.io.netty.handler.codec.http.HttpHeaders.Names.SET_COOKIE;
import static play.shaded.ahc.io.netty.handler.codec.http.HttpHeaders.Names.SET_COOKIE2;
import static play.shaded.ahc.org.asynchttpclient.util.MiscUtils.isNonEmpty;

interface CookieBuilder {

    default List<WSCookie> buildCookies(Map<String, List<String>> headers) {
        List<String> setCookieHeaders = headers.get(SET_COOKIE2);

        if (!isNonEmpty(setCookieHeaders)) {
            setCookieHeaders = headers.get(SET_COOKIE);
        }

        if (isNonEmpty(setCookieHeaders)) {
            List<WSCookie> cookies = new ArrayList<>(setCookieHeaders.size());
            for (String value : setCookieHeaders) {
                Cookie c = isUseLaxCookieEncoder() ? ClientCookieDecoder.LAX.decode(value) : ClientCookieDecoder.STRICT.decode(value);
                if (c != null) {
                    WSCookie wsCookie = new WSCookieBuilder()
                            .setName(c.name())
                            .setValue(c.value())
                            .setDomain(c.domain())
                            .setPath(c.path())
                            .setMaxAge(c.maxAge())
                            .setSecure(c.isSecure())
                            .setHttpOnly(c.isHttpOnly())
                            .build();
                    cookies.add(wsCookie);
                }
            }
            return Collections.unmodifiableList(cookies);
        }

        return Collections.emptyList();
    }

    boolean isUseLaxCookieEncoder();
}
